balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+			b.cc.UpdateState(balancer.State{
+				ConnectivityState: connectivity.TransientFailure,
+				Picker:            &picker{err: fmt.Errorf("failed to create a new subConn: %v", err)},
+			})
+			return
+		}
+		b.subConns.Set(curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressList.increment() {
+			b.endFirstPass(scd.lastErr)
+			return
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Errorf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, newState balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	if newState.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if newState.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		if !b.addressList.seekTo(sd.addr) {
+			// This should not fail as we should have only one subconn after
+			// entering READY. The subconn should be present in the addressList.
+			b.logger.Errorf("Address %q not found address list in  %v", sd.addr, b.addressList.addresses)
+			return
+		}
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If the LB policy is READY, and it receives a subchannel state change,
+	// it means that the READY subchannel has failed.
+	if b.state == connectivity.Ready && newState.ConnectivityState != connectivity.Ready {
+		// Once a transport fails, the balancer enters IDLE and starts from
+		// the first address when the picker is used.
+		b.state = connectivity.Idle
+		b.addressList.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch newState.ConnectivityState {
+		case connectivity.Connecting:
+			// The balancer can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If it's in TRANSIENT_FAILURE, stay in TRANSIENT_FAILURE until
+			// it's READY. See A62.
+			// If the balancer is already in CONNECTING, no update is needed.
+			if b.state == connectivity.Idle {
+				b.state = connectivity.Connecting
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			sd.lastErr = newState.ConnectionError
```

Comment:
> How does requestConnection get triggered and why is it able to encounter a TF subchannel and set an erroring picker?

In `reconcileSubConns()`, the balancer attempts to re-use subconns created for a previous endpoint list produced by the resolver. A subconn could enter backoff and be retained after a resolver update. In such a case, the subconn could still be in backoff when `requestConnection()` passes over it in the `addressList`.

> Why do we need to store this in a field?

Since `requestConnection` can also call `endFirstPass`, it needs a valid error to report to the channel along with the TF update. That's why I decided to store the error.

Reply:
So this is a consequence of getting a new address list and having some addresses already with subchannels that have errored?